### ISSUE TITLE
itk.LabelImageToStatisticsLabelMapFilter issue with median measurements in 16bit images in Python

### ISSUE THREAD

**Issue by jluethi** [InsightSoftwareConsortium/ITK#2482]
### Description

Dear ITK Team
Thanks a lot for making this great library accessible from Python!
We're trying to use ITK in a 3D image processing workflow and use it to make measurements on single cells. For that purpose, one of the things we use is `itk.LabelImageToStatisticsLabelMapFilter` to make the measurements.
When using this with label images and 16bit intensity images, the median measurements have an unexpected behavior. Typically, most of our intensity values are < 1000, but our images sometimes contain some very bright pixels (either in the area to be measured or outside of it. It appears that the GetMedian measurement in this function is doing an 8bit rescaling of the 16bit image, thus being very sensitive to bright/hot pixels.
Is this an intended default? It's very unexpected and makes median measurements unusable for us, as they're just wrong now for the values we're interested in.
I'm providing a dummy-case below to show this issue. In that dummy case, I change 1 pixel value outside the region that is measured to a very high value and compare the measurements.

### Steps to Reproduce

```
# Python test case
import numpy as np
import itk
import copy

def calculate_features(lbl_img, int_img):
    filt = itk.LabelImageToStatisticsLabelMapFilter[
                type(lbl_img), type(int_img), itk.LabelMap.x3
            ].New()
    filt.SetInput1(lbl_img)
    filt.SetInput2(int_img)
    filt.SetComputeHistogram(False)
    filt.SetComputeFeretDiameter(True)
    filt.SetComputePerimeter(True)
    filt.Update()
    features = filt.GetOutput()
    df2 = dict()
    df2['mean'] = np.array([features[l].GetMean() for l in features.GetLabels()])
    df2['median'] = np.array([features[l].GetMedian() for l in features.GetLabels()])
    print('Means: {}'.format(df2['mean']))
    print('Medians: {}'.format(df2['median']))
    #return df2

shape = (50, 50, 1)
# Define a dummy label image
lbl_img_np = np.zeros(shape).astype('uint16')
lbl_img_np[5:10, 5:10, 0] = 1
lbl_img_np[15:20, 5:10, 0] = 2
lbl_img_np[25:30, 5:10, 0] = 3
lbl_img_np[5:10, 15:20, 0] = 4
lbl_img_np[15:20, 15:20, 0] = 5
lbl_img_np[25:30, 15:20, 0] = 6

# Define a dummy intensity image
int_img_np = np.random.randint(500, size=shape).astype('uint16')

# Set some values very high to see if that causes the bug to reproduce
int_img_np_outlier = copy.deepcopy(int_img_np)
int_img_np_outlier[1, 1, 0] = 62000
int_img_np[1, 1, 0] = 100

lbl_img = itk.GetImageFromArray(lbl_img_np)
int_img = itk.GetImageFromArray(int_img_np)
int_img_outlier = itk.GetImageFromArray(int_img_np_outlier)

print('Regular image')
calculate_features(lbl_img, int_img)

print('Image with 1 outlier pixel (outside the measurement area)')
calculate_features(lbl_img, int_img_outlier)
```

### Expected behavior

The code should generate the following output, the same for both test cases:
> Regular image
> Means: [246.08 180.28 251.68 273.52 248.56 268.28]
> Medians: [173.48046875 157.88671875 243.65234375 263.14453125 231.95703125 274.83984375]
> Image with 1 outlier pixel (outside the measurement area)
> Means: [246.08 180.28 251.68 273.52 248.56 268.28]
> Medians: [173.48046875 157.88671875 243.65234375 263.14453125 231.95703125 274.83984375]

### Actual behavior

```none
```
This code actually returns the following output:

> Regular image
> Means: [246.08 180.28 251.68 273.52 248.56 268.28]
> Medians: [173.48046875 157.88671875 243.65234375 263.14453125 231.95703125 274.83984375]
> Image with 1 outlier pixel (outside the measurement area)
> Means: [246.08 180.28 251.68 273.52 248.56 268.28]
> Medians: [242.1875 242.1875 242.1875 242.1875 242.1875 242.1875]

Thus, the Median values are suddenly all the same after I change a pixel outside the measurement area. This is not what I'd expect to happen: Changes in pixels that aren't measured shouldn't influence a measurement.

### Reproducibility

It always reproduces on this test case or images with similarly high pixel values

### Versions

Tested on the python version of itk 5.1.2

### Environment

The test case requires numpy and itk to be installed.

### Additional Information

cc @MaksHess who is working on this with me.

**Comment by thewtex** [comment]
Hi @jluethi ,

Thank you for the report, and excellent description!

@Leengit can you please look into this?

**Comment by Leengit** [comment]
@jluethi, Pull Request #2503 submits a fix, but it is not exactly as you wrote for the expected behavior.  Both images are returning the same median values now, but those values are the actual medians, not an approximation based upon the histogram.  Specifically, I modified your code by adding `np.random.seed(20210413)` before the line that uses `np.random`.  With that, unmodified ITK gives:
```plaintext
Regular image
Means: [254.48 257.   232.56 239.16 258.2  224.36]
Medians: [235.85546875 220.26171875 169.58203125 189.07421875 200.76953125
 126.69921875]
Image with 1 outlier pixel (outside the measurement area)
Means: [254.48 257.   232.56 239.16 258.2  224.36]
Medians: [242.1875 242.1875 242.1875 242.1875 242.1875 242.1875]
```
... and the pull request gives:
```plaintext
Regular image
Means: [254.48 257.   232.56 239.16 258.2  224.36]
Medians: [267. 231. 172. 203. 209. 237.]
Image with 1 outlier pixel (outside the measurement area)
Means: [254.48 257.   232.56 239.16 258.2  224.36]
Medians: [267. 231. 172. 203. 209. 237.]
```
Is this good?

**Comment by jluethi** [comment]
@Leengit Yes, that looks great! Very cool that you could fix this bug! Given how you describe the change you implemented and how it performs on this test case, that should solve this issue! Looking forward to trying it out on my larger datasets again once this is released.
Python itk releases come about once a month, right? Any idea when this will propagate to pypi?

**Comment by dzenanz** [comment]
This fix should be also merged into release branch, so it makes it into the next patch release which will probably come sooner than next minor release.

**Comment by blowekamp** [comment]
I am not in favor of this change. This change creates a whole another array for the image, doubling (?) the memory requirements of the algorithm.

The solution to the users problem for 16-bit images is to set the number of histogram bits to 2^16. For 16-bit pixel types this could be changed to the default.

Also the current behavior should be better documented.

**Comment by blowekamp** [comment]
> This fix should be also merged into release branch, so it makes it into the next patch release which will probably come sooner than next minor release.

This is a significant algorithmic change of established behavior that has existed in ITK for a long time. I have have write usage if this filter many time with the understanding of the current code.

**Comment by jluethi** [comment]
> > This fix should be also merged into release branch, so it makes it into the next patch release which will probably come sooner than next minor release.
> 
> This is a significant algorithmic change of established behavior that has existed in ITK for a long time. I have have write usage if this filter many time with the understanding of the current code.

@blowekamp I don't want to involve myself in the question of optimal implementation, I'm sure you & @Leengit know best about this. 
But I'd highly recommend finding one that actually calculates median values correctly in 16-bit images independent of the image range. This will either result in basically the same result as before, if the image range wasn't too wide, or in much more accurate median measurements in cases like the one described above. In the cases where the outcomes change, it is removing a bias in the measurement and ensuring it actually measures what the name says it would measure.

**Comment by Leengit** [comment]
I could add a flag for the filter.  The default behavior when the flag is not explicitly set by the user would be the values that ITK currently returns, but if the user sets the flag appropriately then the new behavior would be that of this pull request.  We could name the flag `MedianFromHistogram` with `true` as its default, or `ExactMedian` with `false` as its default, or ...?

**Comment by blowekamp** [comment]
> I could add a flag for the filter. The default behavior when the flag is not explicitly set by the user would be the values that ITK currently returns, but if the user sets the flag appropriately then the new behavior would be that of this pull request. We could name the flag `MedianFromHistogram` with `true` as its default, or `ExactMedian` with `false` as its default, or ...?

In my review of the PR, I proposed using the current "ComputeHistogram" flag, or just changed the default number of bins 10 2^16 for 16-bit images.

**Comment by jluethi** [comment]
> I could add a flag for the filter. The default behavior when the flag is not explicitly set by the user would be the values that ITK currently returns, but if the user sets the flag appropriately then the new behavior would be that of this pull request. We could name the flag `MedianFromHistogram` with `true` as its default, or `ExactMedian` with `false` as its default, or ...?

It would be great if by default, this function returns accurate median measurements for 16 bit images. Of course I'd prefer a flag to allow this to not having an option, but I wasn't even expecting the median measurements to be potentially biased in that way until I ran into issues in my data analysis later and probably wouldn't have discovered that I should be setting that flag.

**Comment by Leengit** [comment]
Another trick we have in our bag is to change the exact median calculation to use less memory: a binary search for possible median values would be constant memory.  Its runtime would no longer be _O(n)_ in the number of values; it would be more like _O(n lg n)_ or _O(nb)_ where _b_ is the number of bits in a represented pixel.  Is this space-time trade off worth it in those situations where the user wants the exact median?

**Comment by Leengit** [comment]
The current default number of bins is 128.  If we are going to default to 65536 in the case of `short` pixel values should we also change the default to 256 for `char` pixel values?  And what should we use for `int` `long long`, `float`, and `double`?

To focus discussion: I propose a default of 256 for everything but a `short` and a default of 65536 for a `short`, even though it is a potentially breaking change for all types.  I propose that we have a flag for `MedianFromHistogram` that can be set to `false` by the user to get an exact median by the current implementation for that, which is doubling the memory use relative to unmodified ITK

**Comment by Leengit** [comment]
@jluethi did you try @blowekamp's idea of using `filt.SetNumberOfBins(65536)` before calling `filt.Update()`?  There is still a small change in median between your regular and outlier examples, but it is much smaller.  Does it work for the present until we can figure out what to do long term?
```plaintext
Regular image with default number of bins
Medians: [235.85546875 220.26171875 169.58203125 189.07421875 200.76953125 126.69921875]

Image with 1 outlier pixel (outside the measurement area) with default number of bins
Medians: [242.1875 242.1875 242.1875 242.1875 242.1875 242.1875]

Regular image using filt.SetNumberOfBins(65536)
Medians: [235.99633026 222.00154877 171.00206757 189.00188446 201.00176239 127.99742889]

Image with 1 outlier pixel (outside the measurement area) using filt.SetNumberOfBins(65536)
Medians: [236.03820801 221.84753418 170.7611084  188.73596191 201.0345459 128.18908691]

Exact values without any binning
Medians: [267. 231. 172. 203. 209. 237.]
```

**Comment by Leengit** [comment]
Using `SetNumberOfBins(65536)` gives consistency between the with-outlier and without-outlier cases -- a big win -- but we want both consistency and accuracy and so I am rethinking my above proposal.  Because it could break tests that rely on the present results, I propose that we defer until ITK 5.3 the fix that will return the correct median values.  At that time, tests that are expecting the old values will have to be updated.

And as to the time-space tradeoff, I propose that we allow the user to choose between conserve_time and conserve_memory with ITK 5.3.  The former will be _O(n)_ time and _O(n)_ space and the latter will be _O(nb)_ time and _O(1)_ space where _b_ is the number of bits in the pixel type representation.

**Comment by dzenanz** [comment]
Well, 5.3 is now under development. Deciding the course of action should be done now, and executed soon.

**Comment by jluethi** [comment]
> @jluethi did you try @blowekamp's idea of using `filt.SetNumberOfBins(65536)` before calling `filt.Update()`? There is still a small change in median between your regular and outlier examples, but it is much smaller. Does it work for the present until we can figure out what to do long term?
> 
> ```
> Regular image with default number of bins
> Medians: [235.85546875 220.26171875 169.58203125 189.07421875 200.76953125 126.69921875]
> 
> Image with 1 outlier pixel (outside the measurement area) with default number of bins
> Medians: [242.1875 242.1875 242.1875 242.1875 242.1875 242.1875]
> 
> Regular image using filt.SetNumberOfBins(65536)
> Medians: [235.99633026 222.00154877 171.00206757 189.00188446 201.00176239 127.99742889]
> 
> Image with 1 outlier pixel (outside the measurement area) using filt.SetNumberOfBins(65536)
> Medians: [236.03820801 221.84753418 170.7611084  188.73596191 201.0345459 128.18908691]
> 
> Exact values without any binning
> Medians: [267. 231. 172. 203. 209. 237.]
> ```

Thanks for this idea. Those medians are still regularly off by more than 10% even in this one test example. In my specific use case, we have a background of ~110 that we subtract from measurements. After subtracting the background measurements, the last median above is 18 when it was supposed to be 127. So no, while the error isn't quite as obvious (not all objects get the exact same value), it's still massive and makes median measurements in ITK unreliable for us.

I can appreciate that changing the default behavior comes with complications in test cases. But that median measurement is just simply wrong. It's a rough estimate of a median that is dependent on pixel values outside the area being measured. It is, in my opinion, quite a worrying behavior for a library if by default the measurements can be off by that much. Maybe there is a place for rough estimates of medians and with some restrictions on the data, those estimate may hold reasonably. But I can't have been the first person who got misleading median measurements out of this, it may just be the first time this is noticed or reported...

@dzenanz Regarding the 5.3 release: What kind of timeline is this on? Depending on when this would be released and whether a fix for this can make it into that release, we have to look into switching to a different library like scikit image for our median measurements in the meantime.

**Comment by dzenanz** [comment]
[v5.2.0](https://github.com/InsightSoftwareConsortium/ITK/releases/tag/v5.2.0) was released earlier this month. 5.3 isn't expected to be tagged before July 1st at the earliest, and probably even later. So you will have to use master (after the fix is merged) until then. Or use another library or roll your own median computation.

**Comment by blowekamp** [comment]
>  Those medians are still regularly off by more than 10% even in this one test example. In my specific use case, we have a background of ~110 that we subtract from measurements. After subtracting the background measurements, the last median above is 18 when it was supposed to be 127. So no, while the error isn't quite as obvious (not all objects get the exact same value), it's still massive and makes median measurements in ITK unreliable for us.

When the histogram method is used the maximum expected error for the median should only be 1/2 the value range of a bin. When the number of bins is equal to the number of integers values we would expect an exact median value. For these experiments it looks like there is some error/bug with the current histogram based algorithm.

**Comment by blowekamp** [comment]
@Leengit I added some test to check the current behavior of the histogram based median. It confirmed that the accuracy of the median is 1/2 the value range of a histogram bin. I was unable to repeat the discrepancy is the shared results when the number of bins was increased.

**Comment by Leengit** [comment]
The combination of Pull Requests #2514 (merged) and #2521 (pending) should close this issue.

**Comment by Leengit** [comment]
PR #2521 is merged and this issue can be closed.
